plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Debug/Driver.hs

```diff
@@ -0,0 +1,163 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE GADTs #-}
+module UntypedPlutusCore.Evaluation.Machine.Cek.Debug.Driver
+    ( Breakpoint (..)
+    , Breakpoints
+    , DriverState
+    , Cmd (..)
+    , DTerm
+    , runDriver
+    , handleStep
+    , DebugF (..)
+    -- | Reexport some free functions for convenience
+    , F.MonadFree
+    , F.iterM
+    , F.iterTM
+    , F.partialIterT
+    , F.cutoff
+    ) where
+
+import Annotation
+import UntypedPlutusCore
+import UntypedPlutusCore.Evaluation.Machine.Cek.Debug.Internal
+
+import Data.RandomAccessList.Class qualified as Env
+import Data.Word64Array.Word8 hiding (Index, toList)
+import Control.Monad.Reader
+import Control.Monad.Trans.Free as F
+import Control.Lens hiding (Context)
+import Prettyprinter
+import Data.Ix
+import Data.Set
+import Data.Function
+
+-- | Usually a breakpoint is just a line, but let's be  more generally as a srcspan
+newtype Breakpoint = Breakpoint { unBreakpoint :: SrcSpan }
+    deriving newtype (Show, Pretty, Read, Eq, Ord)
+
+-- | Treat them as `Set`s like `SrcSpan`
+type Breakpoints = Set Breakpoint
+
+-- | The `Term`s that makes sense to debug
+type DTerm uni fun = Term NamedDeBruijn uni fun SrcSpans
+
+-- commands that the debugger can receive fromthe debug-client (tui,cli,test,etc)
+data Cmd
+  =
+  Step -- ^ step the cek once
+  -- big steps
+  | Continue Breakpoints -- ^ continue to the end or until breakpoint reached
+  -- | ContinueIgnore === Continue mempty
+  | Next Breakpoints -- ^ step over aka next
+  | Finish Breakpoints -- ^ run to the end of current function
+  deriving stock (Show, Read)
+
+type DriverState = CekState
+
+-- | The debugger's suspension functor
+data DebugF uni fun a
+  = InputF (Cmd -> a) -- iteratee
+  | LogF String a -- generator
+  | StepF
```

Comment:
Let me explain because I also always forget:

InputF : the driver calls `inputF` and yields/awaits for the client (e.g. TUI) to tell him what the driver should do next (Cmd).

StepF: the driver is about to do one single step. This allows to have different interpretations of the "step" driver action.

When the driver calls `stepF  currentState` it yields its coroutine with its current state (before the stepping interpretation). The *interpreter* of stepF takes the currentState and calculates a newstate from it (e.g. by stepping the cek). The *interpreter* then resumes the driver coroutine with the new state.

```
driver:    stepF currentCekState
interpreter:      newCekState = interp(currentCekState)
interpreter:   resume(driver,newCekState)
```

This layout has 2 benefits:

1) The intepreter can have different interpretations. The logical one is the `handleStep`, a.k.a. the state transition function of the cek state machine. Another easy one is `id`
2) It allows the intrepreter to intercept and  *alter* the `newState` before returning it to the driver and resuming execution.
This is similar to what is done in e.g. C debuggers, where you can pause the execution, alter some variable's value in the stack, and resume the execution with that different value.

Reply:
> Let me explain because I also always forget:

Perfect premise for writing an inline Note :)

Reply:
These benefits though are not taken into advantage anywhere; all the current interpreters (brick,repl,testing) just use the canonical implementation (i.e. handleStep transition function).